### ISSUE TITLE
No easy easy way to write asynchronous request handlers

### ISSUE THREAD

**Issue by socmag** [rejectedsoftware/vibe.d#936]
In NodeJS there is no requirement that the response is written in the request handler....  

That's great, because you can start doing arbitrary async work when the request comes in, and return from the request handler immediately; thus freeing up the event loop and allow processing more client requests... Once the work finally completes , you just do a res.send(...) and finally a res.end() or whatever when you are done.

In vibe.d, I couldn't achieve the same since (looking at the source) the response is cleaned up when the request handler is returned from.  That seems a bit odd, am I missing something?

I did manage to make it work, but had to use a terrible hack.  Since I couldn't return from the request handler, I entered a spin loop that did a yield/sleep until the async work completed.

The vibe.core.yield() was to keep the event loop alive, and the core.thread.sleep(1) to prevent high CPU usage...

Would it be possible to change the semantics please, so that you could continue to use the response object after returning from the request handler.

Here's my awfully ugly hack:

```
router.post("/rest",
    delegate (req, res)
    {
        shared bool waitForComplete = true;
        shared string resultData;

         doTonsOfAsyncWork(req,
              delegate(results)
              {
                  resultData = results;
                  waitForComplete = false;
              }
          );

          while (waitForComplete)
          {
              yield();
              sleep(1);
          }

          res.write(results);
    }
```

what you really want is something like this...

```
router.post("/rest",
    delegate (req, res)
    {
        res.hasAsyncResponse = true;

         doTonsOfAsyncWork(req,
              delegate(results)
              {
                  res.write(results);
                  res.end();
              }
          );
    }
```

I realize I can launch more workers or use the distribute option, but that doesn't really scale very well.  Also, I didn't have much luck with distribute the last time I tried it, but that's another story.

**Comment by socmag** [comment]
Sorry, I'm just catching up after the holidays and hadn't seen the comments thus far....

@etcimon, no it isn't CPU intensive work, it's waiting on lots of RPC calls that are done by a third party library outside the scope of Vibe, all of which have to park the thread.

@s-ludwig

> "Whenever there is any I/O operation waiting for data, the event loop gets control and can schedule other tasks"

Certainly that holds true *assuming* that the I/O is done through Vibe.  However, it still isn't clear to me how to do I/O blocking calls through 3rd party libraries that you do not own, and keep the event loop running.

My original test was quite simple.... I put a long sleep in the request handler, and verified that I could not receive additional requests while the sleep was pending - that's pretty hardcore because library users can't know what calls to external code that they do not control will do internally.

> This is just the general way vibe.d works - it simply frees the programmer from having to manually manage the asynchronous nature of the system.

...at the expense of buying into all the idioms of framework.  It seems like a kitchen sink right now, and there are many incompatibilities and overlap between phobos and Vibe.  It doesn't free up the programmer when they cannot use Vibe for I/O.

Honestly I don't really see the objection to my request as it solves a real world problem

@japplegame the code you provided any sense, as it blocks just the same... okay, if you didn't do the join() then it would work... only it won't work, because the response is cleaned up after the call to the function as I originally stated.

**Comment by japplegame** [comment]
@socmag, can you tell what do you want exactly?
I don't see big difference between hold of request handler and hold of connection. There is no need to release the request handler. Vibe.d allows you to keep thousands of waiting request handlers within single thread without significant overhead.

Sorry for my bad English.

**Comment by dariusc93** [comment]
@socmag If you are trying to perform heavy task, what you could do is create a queue system that uses `runTask` or `async` and submit your request there, and have it return your results later when its done. Thats what I have done in my code though it is tricky to get the task to talk to another task that stores that information off. If you explain what youre trying to get done, we all can better assist you and give you a good answer and possibly an example, but base on your example, it looks like youre trying to have a synchronous task in your request.

**Comment by GeorgeSapkin** [comment]
Here's a modified example from vibe's start page that blocks the handler to simulate some long-running code. While this handler is blocked vibe does not accept any requests. How do I _hold_ the request handler without blocking the event loop and still being able to write to response once the task is done?
```D
import core.thread;
import vibe.d;

shared static this()
{
    auto settings = new HTTPServerSettings;
    settings.port = 8080;
    settings.bindAddresses = ["::0"];
    listenHTTP(settings, &handleRequest);
}

void handleRequest(HTTPServerRequest req, HTTPServerResponse res)
{
    if (req.path == "/")
    {
        // blocking request to expose the issue
        Thread.sleep(5.seconds);
        res.writeBody("Hello, World!", "text/plain");
    }
}
```

Using `runTask` doesn't seem to solve anything as the loop as still blocked and requests don't go through until previous one is complete:
```D
void handleRequest(HTTPServerRequest req, HTTPServerResponse res)
{
    if (req.path == "/")
    {
        runTask({
            // blocking request to expose the issue
            Thread.sleep(5.seconds);
            res.writeBody("Hello, World!", "text/plain");
        });
    }
}
```

Using D `task` to actually do things async fails because as soon as `handleRequest` returns `res` gets cleaned up and a default response is written out:
```
404 - Not Found

Not Found

Internal error information:
Not routes match path '/'
```

```D
import std.parallelism;

void handleRequest(HTTPServerRequest req, HTTPServerResponse res)
{
    if (req.path == "/")
        task!asyncTask(res);
}

void asyncTask(HTTPServerResponse res)
{
    // blocking request to expose the issue
    Thread.sleep(5.seconds);
    res.writeBody("Hello, World!", "text/plain");
}
```

**Comment by dariusc93** [comment]
@GeorgeSapkin as I stated to @socmag, you could push it out into a queue, but overall the Thread.sleep do pause the main thread. You should use `sleep(5.seconds)` instead since

**Comment by socmag** [comment]
@dariusc93 - George showed that example - it's the last one.  The response is cleaned up by the time the work is done - it's dead and buried and can no longer be used.

**Comment by dariusc93** [comment]
@socmag I think I am starting to understand what you are asking now. Give me a bit to create a small example

**Comment by dariusc93** [comment]
@socmag does this work for you

```
import vibe.d;

shared static this()
{
	auto settings = new HTTPServerSettings;
	auto router = new URLRouter;
	settings.port = 8080;
	settings.bindAddresses = ["::1", "127.0.0.1"];
	router.get("/", delegate (req, res){
		task();
		res.writeBody("Hello, World!");
	});

	router.get("/task2", delegate (req, res){
		res.writeBody("Hello, World!");
	});
	listenHTTP(settings, router);

	logInfo("Please open http://127.0.0.1:8080/ in your browser.");
}


bool task()
{
	return async({
		sleep(10.seconds); // simulate some lengthy computation
		return true;
	});
}
```

**Comment by dariusc93** [comment]
Honestly, you dont want to use Thread.sleep as that will pause the main thread. In vibe.d, `sleep` will pause the current operation of that task but should not affect the whole server. Honestly, i do not think the `async` is needed in my example as with just using `sleep(10.seconds)` i am able to access the page on `/task2` without it waiting for the sleep to finish

**Comment by dariusc93** [comment]
and @GeorgeSapkin the example i posted should work for you too

**Comment by socmag** [comment]
@dariusc93

Unfortunately your example doesn't actually use any results from the async task, so it's not very useful.

As for Thread.sleep() - the hint is "Thread".... it blocks the current thread... and it would be pretty unusual for a high performance server to make calls to the HttpRequest handlers on the main thread anyway.  Normally, you have a bunch of workers listening on IOCP that are completely isolated from the main thread.  Don't want to really get into the architecture here though.

Interestingly, we just another test, and instead an HttpServer, we used a TcpServer... and put a Vibe sleep() in the connection callback.  Strangely, the Vibe sleep() did nothing - it didn't sleep at all.

There are multiple issues here, I think George is going to follow on...

**Comment by dariusc93** [comment]
@socmag 

The thing is, `runTask` is not another thread, so using `Thread.sleep` would pause everything (current thread = main thread since runTask is not another thread being spawned according to the code). `sleep` pauses the current task that is running without impacting the other task thats currently active. Vibe.d make used of fibers (co-routines) act like threads, but arent threads, they all are task running on the same thread which is why  `Thread.sleep` pauses everything.

**Comment by socmag** [comment]
Thread.sleep is a synonym for "work"

We just did another experiment where we had both an HttpServer and a TcpServer running.  In the IO callbacks for each we did some "work".

The entire server is throttled to the slowest request, and none of it can run in parallel.

It may work for replying to a client with a simple file, but once you actually do something useful, the whole thing comes to a grinding halt.  I don't mean to sound harsh, but that's the fact.

**Comment by GeorgeSapkin** [comment]
I think we're losing track of the original issue. Let me rephrase that again. I want to call into external, non-vibe API which is blocking and takes a long time to complete and I want to use the result of that call to write a response. While this long task is running I want all my requests to be handled.

**Comment by socmag** [comment]
Yes agreed, we have to have a mechanism in place so that when a request comes in, via whatever protocol, it is possible to process those requests asynchronously **and be able to respond at some point in the future with the result from that work**; otherwise the Vibe event loop will be blocked and not able to process additional requests.

**Comment by dariusc93** [comment]
@GeorgeSapkin could you provide some form of code that is blocking? None of my code is blocking when running inside of a runTask or async and i do have access to external api (the api ive made was written in C but have yet to have time to port it into D but plan on too).

**Comment by japplegame** [comment]
@GeorgeSapkin
> I want to call into external, non-vibe API which is blocking and takes a long time to complete and I want to use the result of that call to write a response.

You can block request handler using several methods. For example, using vibe.core.concurrency`:
```d
import core.thread;
import vibe.d;

shared static this()
{
    auto settings = new HTTPServerSettings;
    settings.port = 8080;
    settings.bindAddresses = ["::0"];
    auto router = new URLRouter;
    router.get("/", &handleRequest);
    listenHTTP(settings, router);
}

import std.stdio;
void handleRequest(HTTPServerRequest req, HTTPServerResponse res)
{
    auto requestTask = Task.getThis();
    // start heavy task in another thread
    new Thread({
        asyncTask(res, requestTask);
    }).start();

    // waiting for signal from async task
    // it does't block processing of other requests
    receiveOnly!bool;
}

void asyncTask(HTTPServerResponse res, Task requestTask)
{
    // do long external task, call C++ libraries, do external HTTP request etc.
    Thread.sleep(10.seconds);
    // write response
    res.writeBody("Hello, World!", "text/plain");
    // send signal to request handler
    requestTask.send(true);
}
```

**Comment by s-ludwig** [comment]
@socmag off the top of my head these are the usual ways to integrate third-party I/O with vibe.d's event loop:

1. A lot of libraries provide a callback based way to customize I/O. OpenSSL's "BIO" is an example of this. These can be used to wrap the vibe.d I/O functions. The only caveat is to not throw any exceptions from the `extern(C)` callbacks.
2. If the library provides access to the underyling file descriptor, it is sometimes possible to set it to non-blocking mode and use [`createFileDescriptorEvent`](http://vibed.org/api/vibe.core.core/createFileDescriptorEvent) to perform the blocking wait before reading/writing to it.
3. If non-blocking mode is not an option, there is [`StdFileStream`](http://vibed.org/api/vibe.stream.stdio/StdFileStream), which transparently uses threads in the background to simulate asynchronous I/O.
4. If nothing else helps, a manual separate thread is the solution. Data can be passed back and forth using for example `vibe.core.concurrency.send`. If a thread pool with one thread per virtual CPU is an acceptable system, `runWorkerTaskH` is another more convenient solution.

**Comment by socmag** [comment]
@japplegame - great, we have a winner!

Yes, we tested that and understand how it works.  that's totally workable, thank you.

@s-ludwig 

I see... unfortunately, of the options available to us, only number (4) appears to be viable, as everything in the library we are calling is highly opaque.

But as I said, that option, and the example @japplegame posted does work.  Now I see what's going on it makes sense

thanks!

It might be nice to add a little demo of this kind of thing.  Very useful indeed!

**Comment by GeorgeSapkin** [comment]
@japplegame, thanks for the example. Using D tasks with `taskPool` instead of an explicit thread seems to work as well.
